rfd/0180-sso-as-mfa-method.md

```diff
@@ -0,0 +1,528 @@
+
+#### First time SSO user
+
+> I am a new Teleport user logging in to the cluster for the first time
+> I want to connect to a resource protected by per-session MFA
+> I have not registered through Teleport, my company uses an IdP provider for login
+
+**Old behavior**
+
+When the user logs in with SSO for the first time, their Teleport user is created
+without any MFA devices registered. In order to access resources protected by
+per-session MFA, they would need to add an MFA device through `tsh mfa add` or
+the settings tab in the WebUI.
+
+Attempting to connect to the cluster without an MFA device currently results in
+an error telling the user to add their first MFA device.
+
+```console
+> tsh ssh server01
+ERROR: MFA is required to access this resource but user has no MFA devices; use 'tsh mfa add' to register MFA devices
+```
+
+**New MFA prompts**
+
+With SSO as an MFA method, this first time user would instead be prompted to
+re-authenticate through their SSO provider for a more seamless experience.
+
+```console
+> tsh ssh server01
+MFA is required to access Node "server01"
+Complete an auth check in your local web browser:
+If browser window does not open automatically, open it by clicking on the link:
+ http://127.0.0.1:60433/f5858c78-75e1-4f3f-b2c5-69e8e76c0ff9
+```
+
+If SSO is not the required MFA method in the cluster, the user will also be
+notified of how to add an MFA device for future MFA checks. The SSO login
+browser will not be opened automatically in order to draw attention to the
+output. The link can still be clicked for easy UX.
+
+```console
+> tsh ssh server01
+MFA is required to access Node "server01"
+Complete an SSO auth check in your local web browser. Open it by clicking on the link:
+ http://127.0.0.1:60433/f5858c78-75e1-4f3f-b2c5-69e8e76c0ff9
+Or register an MFA device with `tsh mfa add` to complete MFA checks with a tap.
+```
+
+Note: the MFA prompt is moved to the end to help it stick out over the SSO link.
+
+Stretch: Rather than having the user add an MFA device with additional steps,
+they should be guided through MFA registration and allowed to complete their
+request afterwards. This would also apply without the SSO MFA option.
+
+```console
+> tsh ssh server01
+MFA is required to access Node "server01"
+Complete an SSO auth check in your local web browser. Open it by clicking on the link:
+ http://127.0.0.1:60433/f5858c78-75e1-4f3f-b2c5-69e8e76c0ff9
+Or register an MFA device to complete MFA checks with a tap.
+Choose device type [WEBAUTHN]:
+
+### User can type to continue the registration process or proceed with SSO re-auth.
+```
+
+#### Existing SSO user with registered MFA method
+
+> I have logged into this cluster before with SSO
+> I have registered one or more MFA devices
+> I want to connect to a resource protected by per-session MFA
+
+**New MFA prompts**
+
+The user will be given the option to pass MFA checks with a registered device
+or with SSO.
```

Comment:
If the user doesn't have the webauthn key plugged in or handy, the SSO option might be preferred, so including the option seems reasonable to me. Checking for a plugged in device is a good idea, but I'll leave it out of scope for now.

Reply:
Checking for plugged is easy with libfido2 and even better with touch ID (as we can check more about the existing key without user interaction). Webauthn.dll is the wildcard - changes are it's doable, but I would have to check.

Reply:
Their primary concern was that the browser would be launched automatically when webauthn was preferred, which is why we'll just print the clickable link.

If we want to introduce preferred MFA methods, I'd rather we allow users to enable it explicitly, such as by marking an MFA method as "preferred" or by adding something like `--mfa-method=otp|webauthn|sso`.

It would look like:
```go
switch mfa-method {
case webauthn:
  if webauthn device connected and webauthn challenge available {
    prompt webauthn only
  }
case sso:
  if sso challenge available {
    prompt sso only
  }
case otp:
  if otp challenge available {
    pompt otp only
  }
}

prompt all available challenges
```